### DISCUSSION TITLE
Analysing BCR sequences

### DISCUSSION THREAD

**Discussion by ayeletperes** [milaboratory/mixcr#1746] [Q&A]
Hey,

I've a three set of processed BCR sequences without primers:

1. Full VDJ region coverage without 5'UTR
2. Full VDJ region coverage with 5'UTR
3. Partial V coverage with full DJ coverage, without 5'UTR

I want to align the sequence to a custom germline library. Is this the correct way to align the sequences?

`
docker run --rm -e MI_LICENSE="${MI_LICENSE}" -v /work/:/work -v /work/mixcr/:/root/.mixcr/libraries/ ghcr.io/milaboratory/mixcr/mixcr:4.6.0 mixcr align -p generic-amplicon -s human -b database --rna --rigid-left-alignment-boundary --rigid-right-alignment-boundary J --keep-non-CDR3-alignments  -OvParameters.geneFeatureToAlign="VRegion"
`

example sequences:

\>Seq_1_set1
CAGGTGCAGCTGGCCCACTCTCGGACTCACATTAAGAACCCTGTTGCCTCAGGGAAGGTTTCCTCCAAGGCTACCGGTTACTGCTTCACTAGTTATGCTAGGAACTGGGTGCGAAAGGCCCGTGGACAAGGGGTTGTATGGATGGGATGGATCAACAACTACAAAGCGAAACCATCGTATGCACAGGGGTTCACCGGACTGAGCGTCTTCTCCTCGCACTCCTCTGTCAGCACGGCATATCTGCAGATCACCAACCTAAACGGTCAGGACACAGCCGTGTGTTACTGTGAGAGAGATTTTCCGCATCAGCACCGGGTCTACTACTACTACAATAAGCACACCTGGGGCCAAGGGACCATGGTCTGCATCTCCTCTT
\>Seq_2_set2
CTGATTTCTGACAGCAATGTGATTCTCTGCGCATCACGACCAATCTGTCCCGGCTGAGGAGCGGCTGGTCGAGTCTGGAGGAGCCTGGGTCTAGCCGGGGGGGTCCCTGAGACTCTCCAGTAAAGGCGCTGGTTTCACCGTTAGTGGCGATGGGATGACCGGGGTCCGCCAGGCGCCACGGAAAGGGTTGGAATGGGTAGCCAACATAAAACAAGTCGGAAGTGTGAAATACTATGTGGACTCTGTGAAGGGACGATTGGACATCTACCAGAAAGAACCTCGCGGACTCACAGTAGCTCCAAATGAACAGCCTGAGAGCCGGGGAAACGGCTGTGATTAACTGTGCGAGCCTTCACCCTTCTGTCTATAGTGGCTCCGTGGGGCTACCTACGGCACTAGCGCCCGGGACCCTGGTCACTATCTCCTAAG
\>Seq_3_set3
CTCTGGGTTTTCCCTCAGCACTAGTGGAATCCGTGTGAGCTGGNTCCGTCAGCCCCCAGGTCAGACCNGGGAGTGGCTTGCACGCANTGACTGGGATGATAATAAAGTCTACACCACATCTCTGAGNACCAGGCTTACCAACTCCGATGACACCTGCNAAAACCAGGTGCTCCTGAGAAGGAGCACCATGAACCCTGTGGACACAGCGACGTATNACTGTGCACGGACCCCCTAGACCCAAGCCATGTGCCGGTCACTGTAGTCGCCGCTTAGGTTCGACCCCTGGGGCCAGGGAACCCTGCTCACCGTCTCCTCGT

**Comment by mizraelson** [comment]
Hi,
Do you want to simply align the sequences and export the alignments for each read, or would you like to assemble the alignments into clonotypes? Also, have you already built the custom library for MiXCR?

**Comment by ayeletperes** [comment]
Hey,

I want to simply align the sequences and export the alignment. I don't need to assemble them into clones. 

I created a custom library with the new AIRR-C reference set.

**Reply by mizraelson** [reply]
Then the command is correct (if you don't need 5'UTR and Leader sequence in the alignments). After generating the `vdjca` file you can use [`mixcr exportAlignemtns'](https://mixcr.com/mixcr/reference/mixcr-export/#alignments) to get a tsv table with alignments for each read.

**Reply by ayeletperes** [reply]
Thanks!

I've compared the run of the 4.6.0 version with the new version of 4.7.0, and it now requires to use the flag of `--assemble-clonotypes-by CDR3` does it have any impact of the resulting alignment? Also now I get just one hit for the V alignment compared to up to 3 in the previous version.

`docker run ghcr.io/milaboratory/mixcr/mixcr:4.7.0 mixcr align -p generic-amplicon --assemble-clonotypes-by CDR3 -s human -b database --rna --keep-non-CDR3-alignments --floating-left-alignment-boundary --rigid-right-alignment-boundary C --write-all -f tmp.fasta tmp.vdjca`

`docker run ghcr.io/milaboratory/mixcr/mixcr:4.7.0 mixcr exportAirr tmp.vdjca tmp.tsv`

**Reply by mizraelson** [reply]
The `--assemble-clonotypes-by CDR3` flag does not affect the align command output. The change in the number of hits is expected as we now export a single `-vHit` instead of all `-vHits` by default. That is to make the `exportAirr` output more compatible with other softwares.

**Reply by ayeletperes** [reply]
Hey,

I have a follow up question about the productive call. I've ran this sequences:

`CAGGTCCAGCTGGTGCAGTCTGGGGCTGAGGAGAAGAAGCCTGGGTCCTCGGTGAAGGTCTCCTGCAAGGCTTCTGGAGGCACCTTCAGCAGCTATGCTATCAGCTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGAAGGATCAACCCTATCTTTGGTACAGCAAACTACGCACAGAAGTTGCAGGGCAGAGTTACGATTACCGCGGACGAATCCACGAGCACAGTCTACATGGAGCTGAGCAGCCTGAGATCTCAGGACACGGCCGTGTATTACTGTGCTGCAGTCAGGCCCTACTGGAACTACTACTACGGTATGGATGTCTGGGGCCAATGGGCCAAGGTCACCGTCTCTTCAG`

IgBlast has determined it productive but MiXCR did not, they both called for the same V, D, and J alleles. What can cause this?

**Reply by mizraelson** [reply]
Hi, can you please share the output where MiXCR calls this as an unproductive sequence? And what command did you use?

**Reply by ayeletperes** [reply]
This is the output

<!DOCTYPE html>

<html>
<head>
	
	

	
</head>

<body>

sequence_id | sequence | rev_comp | productive | v_call | d_call | j_call | c_call | sequence_alignment | germline_alignment | complete_vdj | junction | junction_aa | np1 | np2 | cdr1 | cdr1_aa | cdr2 | cdr2_aa | cdr3 | cdr3_aa | fwr1 | fwr1_aa | fwr2 | fwr2_aa | fwr3 | fwr3_aa | fwr4 | fwr4_aa | v_score | v_cigar | d_score | d_cigar | j_score | j_cigar | c_score | c_cigar | junction_length | np1_length | np2_length | v_germline_start | v_sequence_start | v_germline_end | v_sequence_end | d_germline_start | d_sequence_start | d_germline_end | d_sequence_end | j_germline_start | j_sequence_start | j_germline_end | j_sequence_end | c_germline_start | c_sequence_start | c_germline_end | c_sequence_end | v_alignment_start | v_alignment_end | d_alignment_start | d_alignment_end | j_alignment_start | j_alignment_end | c_alignment_start | c_alignment_end
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
read.0 | CAGGTCCAGCTGGTGCAGTCTGGGGCTGAGGAGAAGAAGCCTGGGTCCTCGGTGAAGGTCTCCTGCAAGGCTTCTGGAGGCACCTTCAGCAGCTATGCTATCAGCTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGAAGGATCAACCCTATCTTTGGTACAGCAAACTACGCACAGAAGTTGCAGGGCAGAGTTACGATTACCGCGGACGAATCCACGAGCACAGTCTACATGGAGCTGAGCAGCCTGAGATCTCAGGACACGGCCGTGTATTACTGTGCTGCAGTCAGGCCCTACTGGAACTACTACTACGGTATGGATGTCTGGGGCCAATGGGCCAAGGTCACCGTCTCTTCAG | F | F | IGHVF6-G21*07 | IGHD2-21*02 | IGHJ6*02 |   | CAGGTCCAGCTGGTGCAGTCTGGGGCTGAGGAGAAGAAGCCTGGGTCCTCGGTGAAGGTCTCCTGCAAGGCTTCTGGAGGCACCTTCAGCAGCTATGCTATCAGCTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGAAGGATCAACCCTATCTTTGGTACAGCAAACTACGCACAGAAGTTGCAGGGCAGAGTTACGATTACCGCGGACGAATCCACGAGCACAGTCTACATGGAGCTGAGCAGCCTGAGATCTCAGGACACGGCCGTGTATTACTGTGCTGCAGTCAGGCCCTACTGGAACTACTACTACGGTATG | CAGGTCCAGCTGGTGCAGTCTGGGGCTGAGGTGAAGAAGCCTGGGTCCTCGGTGAAGGTCTCCTGCAAGGCTTCTGGAGGCACCTTCAGCAGCTATGCTATCAGCTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGAAGGATCATCCCTATCTTTGGTACAGCAAACTACGCACAGAAGTTCCAGGGCAGAGTCACGATTACCGCGGACGAATCCACGAGCACAGCCTACATGGAGCTGAGCAGCCTGAGATCTGAGGACACGGCCGTGTATTACTGTGCNGCAGTCANNNNNNNNNNNNACTACTACTACGGTATG | F |   |   | T | GGCCCTACTGGA | GGAGGCACCTTCAGCAGCTATGCT | GGTFSSYA | ATCAACCCTATCTTTGGTACAGCA | INPIFGTA |   |   | CAGGTCCAGCTGGTGCAGTCTGGGGCTGAGGAGAAGAAGCCTGGGTCCTCGGTGAAGGTCTCCTGCAAGGCTTCT | QVQLVQSGAEEKKPGSSVKVSCKAS | ATCAGCTGGGTGCGACAGGCCCCTGGACAAGGGCTTGAGTGGATGGGAAGG | ISWVRQAPGQGLEWMGR | AACTACGCACAGAAGTTGCAGGGCAGAGTTACGATTACCGCGGACGAATCCACGAGCACAGTCTACATGGAGCTGAGCAGCCTGAGATCTCAGGACACGGCCGTGTATTAC | NYAQKLQGRVTITADESTSTVYMELSSLRSQDTAVYY |   |   | 2726 | 31=1X122=1X36=1X11=1X31=1X28=1X25= | 35 | 291S6N7= | 170 | 310S26N17= |   |   |   | 1 | 12 | 1 | 1 | 290 | 290 | 7 | 292 | 13 | 298 | 27 | 311 | 43 | 327 |   |   |   |   | 1 | 290 | 292 | 298 | 311 | 327 |   |  


</body>

</html>


I've attached the report [test.report.json](https://github.com/user-attachments/files/16711400/test.report.json)

These were the commands:

`docker run --rm -e MI_LICENSE="${MI_LICENSE}" -v /work/:/work -v /work/mixcr/:/root/.mixcr/libraries/ ghcr.io/milaboratory/mixcr/mixcr:4.7.0 mixcr align -p generic-amplicon --assemble-clonotypes-by CDR3 -s human -b database --rna --keep-non-CDR3-alignments --floating-left-alignment-boundary --rigid-right-alignment-boundary C --json-report /work/test.report.json --threads 20 --write-all -f /work/test.fasta /work/test.vdjca`

`docker run -e MI_LICENSE="${MI_LICENSE}" -v /work/:/work -v /work/mixcr/:/root/.mixcr/libraries/ ghcr.io/milaboratory/mixcr/mixcr:4.7.0 mixcr exportAirr -f /work/test.vdjca /work/test.tsv`

**Reply by ayeletperes** [reply]
> The `--assemble-clonotypes-by CDR3` flag does not affect the align command output. The change in the number of hits is expected as we now export a single `-vHit` instead of all `-vHits` by default. That is to make the `exportAirr` output more compatible with other softwares.

Is the alignment command perform any correction to the sequences, though they are without UMIs?

**Reply by mizraelson** [reply]
Hi,

Sorry for the delay. We have looked into the issue with this sequence and found a bug that results in labeling this clone as non-functional. We will fix this in the upcoming release.

Please note that the align command does not perform any corrections.

**Reply by ayeletperes** [reply]
Hey,

Thank you for the update. I think I might have found another bug. I've used the same command to run the sequence below, MiXCR has tagged the sequence as productive though in the translation of the cdr2 region (cdr2_aa) there is a stop codon.



<!DOCTYPE html>
<html>
<body>

<table border="1">
  <tr>
    <th>sequence_id</th>
    <th>sequence</th>
    <th>rev_comp</th>
    <th>productive</th>
    <th>v_call</th>
    <th>d_call</th>
    <th>j_call</th>
    <th>c_call</th>
    <th>sequence_alignment</th>
    <th>germline_alignment</th>
    <th>complete_vdj</th>
    <th>junction</th>
    <th>junction_aa</th>
    <th>np1</th>
    <th>np2</th>
    <th>cdr1</th>
    <th>cdr1_aa</th>
    <th>cdr2</th>
    <th>cdr2_aa</th>
    <th>cdr3</th>
    <th>cdr3_aa</th>
    <th>fwr1</th>
    <th>fwr1_aa</th>
    <th>fwr2</th>
    <th>fwr2_aa</th>
    <th>fwr3</th>
    <th>fwr3_aa</th>
    <th>fwr4</th>
    <th>fwr4_aa</th>
    <th>v_score</th>
    <th>v_cigar</th>
    <th>d_score</th>
    <th>d_cigar</th>
    <th>j_score</th>
    <th>j_cigar</th>
    <th>c_score</th>
    <th>c_cigar</th>
    <th>junction_length</th>
    <th>np1_length</th>
    <th>np2_length</th>
    <th>v_germline_start</th>
    <th>v_sequence_start</th>
    <th>v_germline_end</th>
    <th>v_sequence_end</th>
    <th>d_germline_start</th>
    <th>d_sequence_start</th>
    <th>d_germline_end</th>
    <th>d_sequence_end</th>
    <th>j_germline_start</th>
    <th>j_sequence_start</th>
    <th>j_germline_end</th>
    <th>j_sequence_end</th>
    <th>c_germline_start</th>
    <th>c_sequence_start</th>
    <th>c_germline_end</th>
    <th>c_sequence_end</th>
    <th>v_alignment_start</th>
    <th>v_alignment_end</th>
    <th>d_alignment_start</th>
    <th>d_alignment_end</th>
    <th>j_alignment_start</th>
    <th>j_alignment_end</th>
    <th>c_alignment_start</th>
    <th>c_alignment_end</th>
  </tr>
  <tr>
    <td>read.3</td>
    <td>GAGAAGCAGCTTGTGGAGACTGGGCGAGGCTTGGCACAGCCAGCGCGCTCCCTGAGACTCTCCTGTTGACCTTCTGGATTCATCTTGGGTGATTATGCTATGAGCTGGTTACGCCAGGCTTGAGGGAAGGGGCTGGAGTGTGTAAATTTC</td>
    <td>F</td>
    <td>T</td>
    <td>IGHVF8-G30*02</td>
    <td>IGHD1-1*01</td>
    <td>IGHJ6*03</td>
    <td>NaN</td>
    <td>GCAGCTTGTGGAGACTGGGCGAGGCTTGGCACAGCCAGCGCGCTCCCTGAGACTCTCCTGTTGACCTTCTGGATTCATCTTGGGTGATTATGCTATGAGCTGGTTACGCCAGGCTTGAGGGAAGGGGCTGGAGTGTGTAAATTTC</td>
    <td>GCAGCTGGTGGAGTCTGGGGGAGGCTTGGTACAGCCAGGGCGGTCCCTGAGACTCTCCTGTACAGCTTCTGGATTCACCTTTGGTGATTATGCTATGAGCTGGTTCCGCCAGGCTC</td>
    <td>F</td>
    <td>GGTACTAGAGATATCTTCGGAACGACCACTTACATGGACGTCTGG</td>
    <td>GTRDIFGTTTYMDVW</td>
    <td>TATCTTC</td>
    <td>CACT</td>
    <td>GGATTCATCTTGGGTGATTATGCT</td>
    <td>GFILGDYA</td>
    <td>GTTTGAATCAAAGCTTAAGGTGGAACAGCA</td>
    <td>V*IKA*GGTA</td>
    <td>ACTAGAGATATCTTCGGAACGACCACTTACATGGACGTC</td>
    <td>TRDIFGTTTYMDV</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>ATGAGCTGGTTACGCCAGGCTTGAGGGAAGGGGCTGGAGTGTGTAAATTTC</td>
    <td>MSWLRQA*GKGLECVNF</td>
    <td>GAATACGCCTCCTCTGTGTCAGGCAAATACACCATCTCAAGAGATCAAACCCAAAGCATGGCCCATCTGCAAATGAACAGCCTGAAAACCGAGGACACGGCCGTAAATTAC</td>
    <td>EYASSVSGKYTISRDQTQSMAHLQMNSLKTEDTAVNY</td>
    <td>GGTAAACTGATCACACTCACCTTCCCCTCAG</td>
    <td>GKLITLTFPS_</td>
    <td>1860.0</td>
    <td>5S5N6=1X6=1X5=1X9=1X8=1X3=1X18=2X1=1X12=1X3=1X23=1X9=2X18=1X3=2X4=1X2=1X3=1X9=1X5=1X3=1X11=1X1=1X6=2X5=1X2=1X16=1X1=</td>
    <td>40.0</td>
    <td>309S26N8=</td>
    <td>228.0</td>
    <td>321S37N17=1X3=2X2=1X3=2X5=1X2=1X6=</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>45.0</td>
    <td>7.0</td>
    <td>4.0</td>
    <td>6.0</td>
    <td>6.0</td>
    <td>307.0</td>
    <td>307.0</td>
    <td>27.0</td>
    <td>310.0</td>
    <td>34.0</td>
    <td>317.0</td>
    <td>38.0</td>
    <td>322.0</td>
    <td>83.0</td>
    <td>367.0</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>NaN</td>
    <td>1.0</td>
    <td>302.0</td>
    <td>305.0</td>
    <td>312.0</td>
    <td>317.0</td>
    <td>362.0</td>
    <td>NaN</td>
    <td>NaN</td>
  </tr>
</table>

</body>
</html>

**Reply by mizraelson** [reply]
This is a slightly different case. By default, the authentic AIRR format returns a productive/non-productive value based on the CDR3 sequence. To be consistent with the format in the exportAirr command, we used the same logic, so the sequence is marked as functional or non-functional based on the CDR3. In the regular MiXCR output (not AIRR), you can add a similar column by specifying the feature to use. For example:

```
-isProductive VDJRegion
```

This will add a column with a boolean value based on the entire VDJRegion sequence.
